### DISCUSSION TITLE
Weird behaviour when working with enums with explicit types

### DISCUSSION THREAD

**Discussion by BeastLe9enD** [dotnet/csharplang#5684] [Language Ideas]
Lets say we have an enum:
```CS
public enum MyEnum : ulong {
    VALUE_1 = 1 << 30
}
```

In this case, everything is working without problems.

```CS
public enum MyEnum : ulong {
    VALUE_1 = 1 << 30,
    VALUE_2 = 1 << 31
}
```

If we add VALUE_2, its not working anymore. The compiler complains: Cannot convert source type 'int' into target type 'ulong'.

```CS
public enum MyEnum : ulong {
    VALUE_1 = 1 << 30,
    VALUE_2 = (ulong)1 << 31
}
```
Of course we can fix this by casting to ulong, but in my opinion the compiler should infer that the rhs is of type ulong if the enum is of type ulong. What do you think about this?

**Comment by HaloFour** [comment]
I think the issue is that the compiler treats `1 << 31` as a self-contained expression which is of type `int`.  Casting forces the issue, as well as literal suffixes like the following:

```cs
public enum MyEnum : ulong {
    VALUE_1 = 1UL << 30,
    VALUE_2 = 1UL << 31
}
```

I don't believe the compiler uses inference backwards through the expression for target typing.
